File: packages/core/src/runner/corpus.ts
/**
 * Corpus runner — config (globs → contracts) → aggregated findings across a tree
 * of documents, plus a CI-meaningful exit code (C-0003). Library API, not CLI-only:
 * other consumers reuse it in-process rather than shelling out. Reads files and
 * returns data; never owns argv or `process.exit`.
 *
 * Depends on `../core`; never imports from `../cli`. The traversal walks the tree
 * from `cwd`, matches each file against the config's `include`/`exclude` globs
 * (via `picomatch`), and runs the FIRST matching rule's contract over it
 * (first-match for determinism — see `runCorpus`).
 */
import { readdirSync, readFileSync } from "node:fs";
import { resolve, sep } from "node:path";

import picomatch from "picomatch";

import type { Contract, Finding } from "../core/index.js";

/**
 * The directory → contract config. A flat ordered list of rules: each maps an
 * `include` glob set (minus an optional `exclude` set) to the `contract` that
 * governs the matched files. Rule order is significant — a file is validated
 * against the FIRST rule it matches (see `runCorpus`).
 *
 * `name` is an OPTIONAL human label for the rule's contract (e.g. `capability`,
 * `task`), carried purely for the CLI's run summary to render per-contract counts.
 * The runner never routes on it — routing is by glob, and `runCorpus` returns
 * matched counts by index (`RunStats.matchedByRule`). The declarative front-end
 * populates it from a string contract ref; inline contracts leave it unset.
 */
export interface CorpusConfig {
  rules: Array<{ include: string[]; exclude?: string[]; contract: Contract; name?: string }>;
}

/**
 * Per-run counts `runCorpus` returns alongside the findings, for the CLI's run summary.
 * The counts are tallied inside the single walk (no second pass), so they always satisfy:
 * `filesScanned === walkSync(root).length`, `filesMatched === sum(matchedByRule)`,
 * `filesUnmatched === filesScanned − filesMatched`, and `matchedByRule.length === config.rules.length`.
 * A file removed by the global `include`/`exclude` pre-filter, or matching no rule, counts as UNMATCHED.
 */
export interface RunStats {
  /** every file `walkSync` visited under the run root. */
  filesScanned: number;
  /** files routed to a rule (read + validated). */
  filesMatched: number;
  /** `filesScanned − filesMatched` — scanned but not routed (pre-filtered out or matching no rule). */
  filesUnmatched: number;
  /** matched count per rule, parallel to `config.rules` by index. */
  matchedByRule: number[];
}

/**
 * Identity helper that types a config for `markdown-contract.config.{js,mjs}`.
 * Passthrough is correct here — its only job is to attach the `CorpusConfig` type
 * at the call site so authors get completion and checking on their config literal.
 */
export function defineConfig(config: CorpusConfig): CorpusConfig {
  return config;
}

/** A rule with its include/exclude matchers compiled once, ahead of the walk. */
interface CompiledRule {
  include: ReturnType<typeof picomatch>;
  exclude: ReturnType<typeof picomatch> | null;
  contract: Contract;
}

/** picomatch options — `dot` so dotfiles match like any other file. */
const PICOMATCH_OPTS = { dot: true } as const;

/**
 * Compile a glob set into a matcher under the runner's own matching options (`PICOMATCH_OPTS`).
 * This is the single code path `compile` uses for both `include` and `exclude`, so a test that
 * drives it exercises the runner's real semantics — notably that a single `**\/`-prefixed glob
 * spans both the run root and nested files (so configs need only one entry per rule). A test
 * against this helper fails if those semantics ever regress (an `opts` change or a picomatch
 * upgrade), rather than passing against a re-specified `{ dot: true }` in the test.
 */
export function compileMatcher(globs: string[]): ReturnType<typeof picomatch> {
  return picomatch(globs, PICOMATCH_OPTS);
}

function compile(config: CorpusConfig): CompiledRule[] {
  return config.rules.map((r) => ({
    include: compileMatcher(r.include),
    exclude: r.exclude && r.exclude.length > 0 ? compileMatcher(r.exclude) : null,
    contract: r.contract,
  }));
}

/** Normalize a path to POSIX separators so globs match the same on every platform. */
function toPosix(p: string): string {
  return sep === "/" ? p : p.split(sep).join("/");
}

/**
 * Recursively collect every file under `root`, returned as paths RELATIVE to `root`
 * with POSIX separators (so they feed both the glob matchers and `ctx.path`). The
 * walk is deterministic: directory entries are sorted before recursion. Synchronous
 * so `runCorpus` stays a plain synchronous library call.
 */
function walkSync(root: string): string[] {
  const out: string[] = [];
  const recur = (absDir: string, relDir: string): void => {
    const entries = readdirSync(absDir, { withFileTypes: true });
    entries.sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
    for (const entry of entries) {
      const rel = relDir === "" ? entry.name : `${relDir}/${entry.name}`;
      if (entry.isDirectory()) recur(resolve(absDir, entry.name), rel);
      else if (entry.isFile()) out.push(rel);
    }
  };
  recur(root, "");
  return out;
}

/**
 * Run a config across a document tree and aggregate findings, returning a
 * CI-meaningful exit code.
 *
 * Traversal: walk every file under `cwd` (default `process.cwd()`). For each file,
 * find the FIRST rule whose `include` matches and whose `exclude` does not, and
 * validate the file against that rule's contract. First-match (not multi-match) is
 * deliberate: it makes the run deterministic and lets a config place a specific
 * rule ahead of a catch-all without a file being double-reported. A file that
 * matches no rule is skipped.
 *
 * Each file is read and validated via `contract.validate(source, { path })` where
 * `path` is the file's path relative to `cwd` (POSIX-separated); the engine stamps
 * that path onto every finding, so findings already carry their file location and
 * are deterministically sorted within the file.
 *
 * An optional global `include` / `exclude` pre-filter narrows the run *before* rule
 * matching, independent of the config's own per-rule globs: a file is considered only
 * if it matches at least one `include` glob (when any are given) and no `exclude` glob.
 * This is an AND-narrowing (a file must satisfy both the pre-filter and a rule), so it
 * works uniformly across an inline contract and a multi-rule `--config` — something a
 * per-rule glob list can't express. All globs are matched relative to `cwd`.
 *
 * Exit-code policy (AC-2): `0` when no `error`-level finding exists across the whole
 * corpus, `1` when any `error`-level finding is present. `2` is reserved for
 * usage/config errors and is raised by the CLI layer, never here — this function
 * only ever returns `0` or `1`, so the CLI can layer `2` on top.
 */
export function runCorpus(
  config: CorpusConfig,
  opts?: {
    format?: "human" | "json" | "sarif";
    cwd?: string;
    include?: string[];
    exclude?: string[];
  },
): { findings: Finding[]; exitCode: number; stats: RunStats } {
  const root = resolve(opts?.cwd ?? process.cwd());
  const rules = compile(config);

  // Optional global pre-filter, applied before rule matching (AND-narrowing).
  const include =
    opts?.include && opts.include.length > 0 ? picomatch(opts.include, PICOMATCH_OPTS) : null;
  const exclude =
    opts?.exclude && opts.exclude.length > 0 ? picomatch(opts.exclude, PICOMATCH_OPTS) : null;

  const findings: Finding[] = [];
  const files = walkSync(root);

  // Run counts, tallied inside the single walk (no second pass). `matchedByRule` is
  // parallel to `config.rules` by index; the matched rule is found by INDEX so its
  // count can be attributed back to the rule the CLI labels.
  const matchedByRule = new Array<number>(config.rules.length).fill(0);
  let filesMatched = 0;

  for (const rel of files) {
    const posixRel = toPosix(rel);
    if (exclude && exclude(posixRel)) continue;
    if (include && !include(posixRel)) continue;
    const idx = rules.findIndex((r) => r.include(posixRel) && !(r.exclude && r.exclude(posixRel)));
    if (idx === -1) continue;
    matchedByRule[idx] = (matchedByRule[idx] ?? 0) + 1;
    filesMatched += 1;
    const source = readFileSync(resolve(root, rel), "utf8");
    const result = rules[idx]!.contract.validate(source, { path: posixRel });
    findings.push(...result.findings);
  }

  const stats: RunStats = {
    filesScanned: files.length,
    filesMatched,
    filesUnmatched: files.length - filesMatched,
    matchedByRule,
  };

  const hasError = findings.some((f) => f.level === "error");
  return { findings, exitCode: hasError ? 1 : 0, stats };
}

File: crates/markdown-contract-engine/src/runner/corpus.rs
//! Corpus runner — config (globs → contracts) → aggregated findings across a set of
//! documents, plus a CI-meaningful exit code (C-0003), mirroring the TS
//! `runner/corpus.ts`.
//!
//! The CORE is fs-free: [`run_corpus`] takes the file set as data — `(relative posix
//! path, source)` pairs, in walk order — matches each path against the config's
//! `include` / `exclude` globs, and validates the FIRST matching rule's contract over
//! it (first-match for determinism). The std::fs walker that feeds it from a directory
//! tree lives in the thin native layer ([`crate::runner::native`]).
//!
//! Glob semantics replicate picomatch (`{ dot: true }`): `*` never crosses `/`
//! (globset's `literal_separator`), `**/` spans zero or more directories (so one
//! `**/*.md` covers the run root and nested files), dotfiles match like any other file,
//! and `{a,b}` alternation works.
//!
//! Exit-code policy (AC-2): `0` when no `error`-level finding exists, `1` when one
//! does. `2` is reserved for usage/config errors and is raised by the hosting layer
//! (e.g. a `DeclarativeError` from config loading), never here.

use globset::{GlobBuilder, GlobSet, GlobSetBuilder};

use crate::contract::Contract;
use crate::finding::{Finding, FindingLevel};
use crate::validate::validate;

/// One routing rule: an `include` glob set (minus an optional `exclude` set) mapped to
/// the contract that governs the matched files. `name` is an optional human label
/// (populated from a string contract ref by the declarative config loader); the runner
/// never routes on it.
pub struct CorpusRule {
    pub include: Vec<String>,
    pub exclude: Vec<String>,
    pub contract: Contract,
    pub name: Option<String>,
}

/// The directory → contract config: a flat ORDERED list of rules. Rule order is
/// significant — a file is validated against the FIRST rule it matches.
#[derive(Default)]
pub struct CorpusConfig {
    pub rules: Vec<CorpusRule>,
}

/// Per-run counts returned alongside the findings. Always satisfies
/// `files_matched == matched_by_rule.iter().sum()` and
/// `files_unmatched == files_scanned - files_matched`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RunStats {
    /// every file offered to the run
    pub files_scanned: usize,
    /// files routed to a rule (validated)
    pub files_matched: usize,
    /// offered but not routed (pre-filtered out or matching no rule)
    pub files_unmatched: usize,
    /// matched count per rule, parallel to `config.rules` by index
    pub matched_by_rule: Vec<usize>,
}

/// The run outcome: aggregated findings (each already deterministically sorted within
/// its file, stamped with its file path), the exit code, and the run counts.
pub struct RunOutcome {
    pub findings: Vec<Finding>,
    /// `0` = no error-level finding; `1` = at least one
    pub exit_code: i32,
    pub stats: RunStats,
}

/// An optional global pre-filter, applied before rule matching (AND-narrowing): a file
/// is considered only if it matches at least one `include` glob (when any are given)
/// and no `exclude` glob.
#[derive(Default)]
pub struct RunOptions {
    pub include: Vec<String>,
    pub exclude: Vec<String>,
}

/// A glob set that failed to compile — a config error (exit-code-2 territory for the
/// hosting layer).
#[derive(Debug, thiserror::Error)]
#[error("invalid glob '{glob}': {reason}")]
pub struct GlobError {
    pub glob: String,
    pub reason: String,
}

/// Compile a glob list into a matcher under the runner's picomatch-equivalent options:
/// `literal_separator` (a `*` never crosses `/`) — dotfiles already match by default.
pub fn compile_matcher(globs: &[String]) -> Result<GlobSet, GlobError> {
    let mut builder = GlobSetBuilder::new();
    for glob in globs {
        let compiled = GlobBuilder::new(glob)
            .literal_separator(true)
            .build()
            .map_err(|e| GlobError {
                glob: glob.clone(),
                reason: e.to_string(),
            })?;
        builder.add(compiled);
    }
    builder.build().map_err(|e| GlobError {
        glob: globs.join(", "),
        reason: e.to_string(),
    })
}

struct CompiledRule {
    include: GlobSet,
    exclude: Option<GlobSet>,
}

fn compile(config: &CorpusConfig) -> Result<Vec<CompiledRule>, GlobError> {
    config
        .rules
        .iter()
        .map(|r| {
            Ok(CompiledRule {
                include: compile_matcher(&r.include)?,
                exclude: if r.exclude.is_empty() {
                    None
                } else {
                    Some(compile_matcher(&r.exclude)?)
                },
            })
        })
        .collect()
}

/// Index of the FIRST rule whose include matches and exclude does not, or `None`.
fn first_matching_rule(rules: &[CompiledRule], path: &str) -> Option<usize> {
    rules.iter().position(|r| {
        r.include.is_match(path) && !r.exclude.as_ref().is_some_and(|e| e.is_match(path))
    })
}

/// Run a config across a file set and aggregate findings.
///
/// `files` supplies `(relative posix path, source)` pairs in the caller's walk order
/// (the native walker yields the TS runner's deterministic per-directory sort). For
/// each file the FIRST rule whose `include` matches and whose `exclude` does not
/// validates it, with the file's relative path stamped onto every finding. A file that
/// matches no rule is skipped.
pub fn run_corpus<'a>(
    config: &CorpusConfig,
    files: impl IntoIterator<Item = (&'a str, &'a str)>,
    opts: &RunOptions,
) -> Result<RunOutcome, GlobError> {
    let rules = compile(config)?;
    let include = if opts.include.is_empty() {
        None
    } else {
        Some(compile_matcher(&opts.include)?)
    };
    let exclude = if opts.exclude.is_empty() {
        None
    } else {
        Some(compile_matcher(&opts.exclude)?)
    };

    let mut findings = Vec::new();
    let mut matched_by_rule = vec![0usize; config.rules.len()];
    let mut files_scanned = 0usize;
    let mut files_matched = 0usize;

    for (path, source) in files {
        files_scanned += 1;
        if exclude.as_ref().is_some_and(|e| e.is_match(path)) {
            continue;
        }
        if include.as_ref().is_some_and(|i| !i.is_match(path)) {
            continue;
        }
        let Some(idx) = first_matching_rule(&rules, path) else {
            continue;
        };
        matched_by_rule[idx] += 1;
        files_matched += 1;
        findings.extend(validate(source, &config.rules[idx].contract, path));
    }

    let has_error = findings.iter().any(|f| f.level == FindingLevel::Error);
    Ok(RunOutcome {
        findings,
        exit_code: i32::from(has_error),
        stats: RunStats {
            files_scanned,
            files_matched,
            files_unmatched: files_scanned - files_matched,
            matched_by_rule,
        },
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::contract::{Contract, LevelOpts, section, sections};

    fn overview_contract() -> Contract {
        Contract::new().body(sections(LevelOpts::default(), vec![section("Overview")]))
    }

    fn rule(include: &[&str], exclude: &[&str]) -> CorpusRule {
        CorpusRule {
            include: include.iter().map(|s| s.to_string()).collect(),
            exclude: exclude.iter().map(|s| s.to_string()).collect(),
            contract: overview_contract(),
            name: None,
        }
    }

    // Contract first: one rule, two files — the matching file validates, the other is
    // skipped, and the exit code reflects the error.
    #[test]
    fn routes_by_glob_and_returns_a_ci_exit_code() {
        let config = CorpusConfig {
            rules: vec![rule(&["**/*.md"], &[])],
        };
        let files = [
            ("docs/bad.md", "## Wrong\n\nprose\n"),
            ("notes.txt", "not markdown"),
        ];
        let out = run_corpus(&config, files, &RunOptions::default()).unwrap();
        assert_eq!(out.exit_code, 1);
        assert_eq!(out.findings.len(), 1);
        assert_eq!(out.findings[0].path, "docs/bad.md");
        assert_eq!(
            out.stats,
            RunStats {
                files_scanned: 2,
                files_matched: 1,
                files_unmatched: 1,
                matched_by_rule: vec![1],
            }
        );
    }

    #[test]
    fn clean_corpus_exits_zero() {
        let config = CorpusConfig {
            rules: vec![rule(&["**/*.md"], &[])],
        };
        let out = run_corpus(
            &config,
            [("a.md", "## Overview\n\nfine\n")],
            &RunOptions::default(),
        )
        .unwrap();
        assert_eq!(out.exit_code, 0);
        assert_eq!(out.findings, vec![]);
    }

    // First-match routing: the specific rule ahead of the catch-all wins.
    #[test]
    fn first_match_routing_is_by_rule_order() {
        let config = CorpusConfig {
            rules: vec![rule(&["**/A-*.md"], &[]), rule(&["**/*.md"], &[])],
        };
        let files = [
            ("A-1.md", "## Overview\n\nok\n"),
            ("b.md", "## Overview\n\nok\n"),
        ];
        let out = run_corpus(&config, files, &RunOptions::default()).unwrap();
        assert_eq!(out.stats.matched_by_rule, vec![1, 1]);
    }

    #[test]
    fn per_rule_exclude_and_global_pre_filter_narrow() {
        let config = CorpusConfig {
            rules: vec![rule(&["**/*.md"], &["**/skip-*.md"])],
        };
        let out = run_corpus(
            &config,
            [("skip-1.md", "x"), ("keep.md", "## Overview\n\nok\n")],
            &RunOptions::default(),
        )
        .unwrap();
        assert_eq!(out.stats.files_matched, 1);

        let out = run_corpus(
            &config,
            [
                ("keep.md", "## Overview\n\nok\n"),
                ("docs/keep.md", "## Overview\n\nok\n"),
            ],
            &RunOptions {
                include: vec!["docs/**".into()],
                ..Default::default()
            },
        )
        .unwrap();
        assert_eq!(out.stats.files_matched, 1);
    }

    // The picomatch parity cases the TS runner pins: a single `**/`-glob spans the run
    // root and nested files; `*` never crosses a separator; dotfiles match.
    #[test]
    fn glob_semantics_match_picomatch() {
        let m = compile_matcher(&["**/*.md".into()]).unwrap();
        assert!(m.is_match("a.md")); // root level
        assert!(m.is_match("docs/deep/a.md")); // nested
        assert!(m.is_match(".hidden.md")); // dot: true
        let single = compile_matcher(&["docs/*.md".into()]).unwrap();
        assert!(single.is_match("docs/a.md"));
        assert!(!single.is_match("docs/deep/a.md")); // * does not cross /
        let brace = compile_matcher(&["**/{a,b}.md".into()]).unwrap();
        assert!(brace.is_match("x/a.md"));
        assert!(brace.is_match("b.md"));
        assert!(!brace.is_match("c.md"));
    }

    #[test]
    fn an_invalid_glob_is_a_config_error() {
        assert!(compile_matcher(&["a[".into()]).is_err());
    }
}
